tiledb/sm/c_api/tiledb.h

```diff
@@ -2288,6 +2291,36 @@ TILEDB_EXPORT int32_t tiledb_query_alloc(
 TILEDB_EXPORT int32_t tiledb_query_set_subarray(
     tiledb_ctx_t* ctx, tiledb_query_t* query, const void* subarray);
 
+/**
+ * Sets a subarray object to the query.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_subarray_t* subarray;;
+ * // Code that allocates and prepares the subarray.
+ * tiledb_query_set_subarray_2(ctx, query, subarray);
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param query The TileDB query.
+ * @param subarray The subarray to be set.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ *
+ * @note Setting a subarray object to the query does not work with
+ *     writes. This function will output an error if the query was
+ *     created in write mode.
+ *
+ * @note Setting a subarray object to the query does not work for
+ *     dense arrays. This function will output an error if the
+ *     query was created for a dense array.
+ *
+ * @note Setting a subarray object to the query does not work for
+ *     any layout other than `TILEDB_UNORDERED`.
+ */
+TILEDB_EXPORT int32_t tiledb_query_set_subarray_2(
```

Comment:
Suggestions for name:
- `tiledb_query_set_subarray_object()`
- `tiledb_query_set_subarray_advanced()`
- `tiledb_query_set_advanced_subarray()`

Another (breaking) suggestion would be to rename the existing `tiledb_query_set_subarray` to `tiledb_query_set_single_subarray` and then the new API could be `tiledb_query_set_subarray`.

@jakebolewski ?